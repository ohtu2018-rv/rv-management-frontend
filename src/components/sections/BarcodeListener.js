import React, { Component } from 'react';
import { connect } from 'react-redux';
import { SuccessBtn } from './../buttons/Buttons';
import { setProductSelected } from '../../reducers/productReducer';
import { setBarcode, toggleBarcodeVisibility, clearBarcode } from '../../reducers/barcodeListenerReducer';
import { errorMessage } from '../../reducers/notificationReducer';
import { withRouter } from 'react-router-dom';

export class BarcodeListener extends Component {

    handleInputEvent(event) {
        this.props.setBarcode(event.target.value);
    }

    handleSubmit(event) {
        event.preventDefault();
        var product = this.props.products.find(
            prod => prod.product_barcode === this.props.barcode
        );
        if (product) {
            this.props.setProductSelected(product.product_id);
            this.props.history.push(
                `/products/${product.product_id}/stock`
            );
        } else {
            var box = this.props.boxes.find(
                b => b.box_barcode === this.props.barcode
            );
            if (box) {
                this.props.setProductSelected(box.product_id);
                this.props.history.push(
                    `/products/${box.product_id}/box`
                );
            } else {
                if (window.confirm('Not found.\nWant to create a new item?')) {
                    if (window.confirm('Create a box?')) {
                        this.props.history.push('/products/create/box');
                    } else {
                        this.props.history.push('/products/create/product');
                    }
                } else {
                    this.props.errorMessage('Product not found');
                }
            }
        }
        this.props.clearBarcode();
    }

    handleOpen(event) {
        this.props.clearBarcode();
        this.props.toggleBarcodeVisibility(true);
    }

    render() {
        return (
            <React.Fragment>
                {this.props.barcodeVisible && (
                    <form onSubmit={event => this.handleSubmit(event)}>
                        <input
                            id="barcodeInput"
                            style={{
                                paddingLeft: 16,
                                paddingRight: 16,
                                paddingTop: 11,
                                paddingBottom: 11,
                                fontSize: 16,
                                textAlign: 'center',
                                display: 'inlineBlock'
                            }}
                            ref={input => {
                                input && input.focus();
                            }}
                            value={this.props.barcode}
                            placeholder="Lue viivakoodi"
                            pattern="\d*"
                            onChange={event => this.handleInputEvent(event)}
                        />
                        <SuccessBtn
                            onClick={event => this.handleSubmit(event)}
                            type="button"
                        >
                            Hae
                        </SuccessBtn>
                    </form>
                )}
                {this.props.barcodeVisible || (
                    <SuccessBtn onClick={event => this.handleOpen(event)}>
                        Lue viivakoodi
                    </SuccessBtn>
                )}
            </React.Fragment>
        );
    }
}

const mapStateToProps = state => {
    return {
        products: state.product.products,
        boxes: state.box.boxes,
        barcode: state.barcodeListener.barcode,
        barcodeVisible: state.barcodeListener.visible
    };
};

const mapDispatchToProps = {
    setProductSelected,
    errorMessage,
    setBarcode,
    clearBarcode,
    toggleBarcodeVisibility
};

export default withRouter(connect(mapStateToProps, mapDispatchToProps)(BarcodeListener));
