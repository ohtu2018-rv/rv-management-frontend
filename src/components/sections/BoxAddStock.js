import React from 'react';
import { Row, Col } from 'react-flexbox-grid';
import './styles/BoxAddStock.css';
import { connect } from 'react-redux';
import { Redirect, withRouter } from 'react-router-dom';

import { addStock } from '../../reducers/boxReducer';
import { setUpgradeStock } from '../../reducers/productReducer';

export class BoxAddStock extends React.Component {
    constructor(props) {
        super(props);

        this.calculateProductSellprice = this.calculateProductSellprice.bind(
            this
        );
        this.calculateProductSellpriceAndBoxcost = this.calculateProductSellpriceAndBoxcost.bind(
            this
        );
        this.formSubmit = this.formSubmit.bind(this);
    }

    updateFields() {
        if (this.props.box) {
            this.barcodeInput.value = this.props.box.box_barcode;
            this.productBarcodeInput.value = this.props.product.product_barcode;
            this.costInput.value = (this.props.product.buyprice / 100).toFixed(2);
            this.amountInput.value = this.props.box.items_per_box;
            this.marginInput.value = this.props.globalMargin;
            this.calculateProductSellpriceAndBoxcost();
        }
    }

    calculateProductSellprice() {
        const boxcost = parseFloat(this.boxcostInput.value);
        const amount = parseFloat(this.amountInput.value);
        const margin = parseFloat(this.marginInput.value);
        const productcost = boxcost / amount;

        this.costInput.value = productcost.toFixed(2);
        this.sellpriceInput.value = (
            productcost *
            ((100 + margin) / 100)
        ).toFixed(2);
    }

    calculateProductSellpriceAndBoxcost() {
        const productcost = parseFloat(this.costInput.value);
        const margin = parseFloat(this.marginInput.value);
        const amount = parseFloat(this.amountInput.value);

        this.boxcostInput.value = (productcost * amount).toFixed(2);
        this.sellpriceInput.value = (
            productcost *
            ((100 + margin) / 100)
        ).toFixed(2);
    }

    componentDidMount() {
        this.updateFields();
    }

    componentDidUpdate() {
        if (!this.props.upgradeStock) {
            this.updateFields();
        } else {
            const link = '/products/' + this.props.product.product_id;
            this.props.history.push(link);
        }
    }

    componentWillUnmount() {}

    formSubmit(event) {
        event.preventDefault();

        const box = {
            product_id: this.props.product.product_id,
            sellprice: this.sellpriceInput.value,
            buyprice: this.costInput.value,
            boxes: this.quantityInput.value
        };

        this.props.addStock(this.props.box.box_barcode, box, this.props.token);
    }

    render() {
        return (
            <div className="box-stocking-form">
                <form onSubmit={this.formSubmit}>
                    <Row>
                        <Col xs={3}>
                            <label htmlFor="barcode">Laatikon viivakoodi</label>
                        </Col>
                        <Col xs={9}>
                            <input
                                id="barcode"
                                name="barcode"
                                type="text"
                                ref={input => {
                                    this.barcodeInput = input;
                                }}
                            />
                        </Col>
                    </Row>
                    <Row>
                        <Col xs={3}>
                            <label htmlFor="cost">
                                Laatikon sisäänostohinta
                            </label>
                        </Col>
                        <Col xs={9}>
                            <input
                                id="boxcost"
                                name="boxcost"
                                type="number"
                                step="0.01"
                                min="0"
                                defaultValue="1.50"
                                ref={input => {
                                    this.boxcostInput = input;
                                }}
                                onChange={this.calculateProductSellprice}
                            />
                            <span className="unit">&euro;</span>
                        </Col>
                    </Row>
                    <Row>
                        <Col xs={3}>
                            <label htmlFor="margin">Kate</label>
                        </Col>
                        <Col xs={9}>
                            <input
                                id="margin"
                                name="margin"
                                type="number"
                                step="1"
                                min="0"
                                defaultValue={this.props.globalMargin}
                                ref={input => {
                                    this.marginInput = input;
                                }}
                                onChange={this.calculateProductSellprice}
                            />
                            <span className="unit">%</span>
                        </Col>
                    </Row>
                    <Row>
                        <Col xs={3}>
                            <label htmlFor="amount">Tuotteita laatikossa</label>
                        </Col>
                        <Col xs={9}>
                            <input
                                id="amount"
                                name="amount"
                                type="number"
                                step="1"
                                defaultValue="10"
                                ref={input => {
                                    this.amountInput = input;
                                }}
                                disabled
                            />
                            <span className="unit">kpl</span>
                        </Col>
                    </Row>
                    <Row>
                        <Col xs={3}>
                            <label htmlFor="quantity">Laatikoiden määrä</label>
                        </Col>
                        <Col xs={9}>
                            <input
                                id="quantity"
                                name="quantity"
                                type="number"
                                defaultValue="1"
                                min="1"
                                step="1"
                                ref={input => {
                                    this.quantityInput = input;
                                }}
                            />
                            <span className="unit">kpl</span>
                        </Col>
                    </Row>
                    <div className="product-wrapper">
                        <Row>
                            <Col xs={3}>
                                <label className="product-title">
                                    Laatikko sisältää:
                                </label>
                            </Col>
                            <Col xs={9}>
                                <label className="product-title">
                                    {this.props.product.product_name}
                                </label>
                            </Col>
                        </Row>
                        <Row>
                            <Col xs={3}>
                                <label htmlFor="barcode">
                                    Tuotteen viivakoodi
                                </label>
                            </Col>
                            <Col xs={9}>
                                <input
                                    id="barcode"
                                    name="barcode"
                                    type="text"
                                    ref={input => {
                                        this.productBarcodeInput = input;
                                    }}
                                />
                            </Col>
                        </Row>
                        <Row>
                            <Col xs={3}>
                                <label htmlFor="cost">
                                    Tuotteen sisäänostohinta
                                </label>
                            </Col>
                            <Col xs={9}>
                                <input
                                    id="cost"
                                    name="cost"
                                    type="number"
                                    step="0.01"
                                    min="0"
                                    defaultValue="1.50"
                                    ref={input => {
                                        this.costInput = input;
                                    }}
                                    onChange={
                                        this.calculateProductSellpriceAndBoxcost
                                    }
                                />
                                <span className="unit">&euro;</span>
                            </Col>
                        </Row>
                        <Row>
                            <Col xs={3}>
                                <label htmlFor="sellprice">
                                    Tuotteen myyntihinta
                                </label>
                            </Col>
                            <Col xs={9}>
                                <input
                                    id="sellprice"
                                    name="sellprice"
                                    type="number"
                                    step="0.01"
                                    ref={input => {
                                        this.sellpriceInput = input;
                                    }}
                                />
                                <span className="unit">&euro;</span>
                            </Col>
                        </Row>
                    </div>
                    <Row>
                        <Col xs={6}>
                            <input
                                type="submit"
                                className="btn btn-success"
                                value="Osta sisään"
                            />
                        </Col>
                    </Row>
                </form>
            </div>
        );
    }
}

const mapDispatchToProps = {
    addStock,
    setUpgradeStock
};

const mapStateToProps = state => {
    return {
        globalMargin: state.product.globalMargin,
        token: state.authentication.accessToken,
        upgradeStock: state.product.upgradeStock
    };
};

export default withRouter(connect(mapStateToProps, mapDispatchToProps)(BoxAddStock));
