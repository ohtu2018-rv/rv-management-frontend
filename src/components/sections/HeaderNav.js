import React, { Component } from 'react';
import { NavLink } from 'react-router-dom';
import './styles/HeaderNav.css';

const activeClass = 'header-nav-active';

export class HeaderNav extends Component {
    render() {
        return (
            <div className="header-nav">
                <ul>
                    <li>
                        <NavLink to="/products" activeClassName={activeClass}>Tuotteet</NavLink>
                    </li>
                    <li>
                        <NavLink to="/users" activeClassName={activeClass}>Käyttäjät</NavLink>
                    </li>
                    <li>
                        <NavLink to="/reports" activeClassName={activeClass}>Raportit</NavLink>
                    </li>
                </ul>
            </div>
        );
    }
}

export default HeaderNav;
