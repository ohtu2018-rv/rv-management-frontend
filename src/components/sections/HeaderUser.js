import React, { Component } from 'react';
import { connect } from 'react-redux';
import { DangerBtn } from '../buttons/Buttons';
import './styles/HeaderUser.css';

class HeaderUser extends Component {
    render() {
        return (
            <div className="header-user">
                <DangerBtn small>Kirjaudu ulos</DangerBtn>
            </div>
        );
    }
}

const mapDispatchToProps = state => {
    return {
        isAuthenticated: state.authentication.isAuthenticated
    };
};

export default HeaderUser;
