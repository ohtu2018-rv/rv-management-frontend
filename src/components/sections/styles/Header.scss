@import 'vars';

.topheader {
  display: flex;
  width: 100%;
  background: $header-background;
  box-shadow: 0 0 4px 2px rgba(0, 0, 0, 0.3);

  h1 {
    display: inline-block;
    padding: 19px 25px 18px 20px;
    font-size: 1.7em;
    color: #eee;
    font-weight: normal;
    vertical-align: middle;
  }

  .user-info {
    margin-left: auto;
  }
}
