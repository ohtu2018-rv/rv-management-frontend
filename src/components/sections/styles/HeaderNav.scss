.header-nav {
  ul {
    display: flex;
    padding: 0;
    margin: 0;

    li {
      display: block;
    }

    li a {
      display: block;
      padding: 22px 16px;
      color: #aaa;
      text-decoration: none;
    }

    li a.header-nav-active {
      background: #000;
      color: #fff;
    }
  }
}
