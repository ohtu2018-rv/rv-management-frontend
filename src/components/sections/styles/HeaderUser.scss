@import "vars";

.header-user {
  margin-left: auto;
  color: #fff;
  padding: 14px 20px 13px 20px;
}
