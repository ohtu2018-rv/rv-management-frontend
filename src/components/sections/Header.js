import React, { Component } from 'react';
import { connect } from 'react-redux';
import HeaderNav from './HeaderNav';
import HeaderUser from './HeaderUser';
import { withRouter } from 'react-router-dom';

import './styles/Header.css';

export class Header extends Component {
    render() {
        return (
            <header className="topheader">
                <div className="header-title">
                    <h1>RV management</h1>
                </div>
                { this.props.isAuthenticated && <HeaderNav/> }
                { this.props.isAuthenticated && <HeaderUser/> }
            </header>
        );
    }
}

const mapStateToProps = state => {
    return {
        isAuthenticated: state.authentication.isAuthenticated
    };
};

export default withRouter(connect(mapStateToProps)(Header));
