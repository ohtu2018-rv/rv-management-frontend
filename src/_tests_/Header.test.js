import React from 'react';
import { Header } from '../components/sections/Header';
import { mount, shallow } from 'enzyme';
import { HeaderNav } from '../components/sections/HeaderNav';
import HeaderUser from '../components/sections/HeaderUser';
import renderer from 'react-test-renderer';

it('renders without crashing', () => {
    mount(<Header />);
    const header = renderer.create(<Header />).toJSON();
    expect(header).toMatchSnapshot();
});

it('does not render navigation and user info when not logged in', () => {
    const header = mount(<Header />);
    expect(header.find(HeaderNav).length).toBe(0);
    expect(header.find(HeaderUser).length).toBe(0);

    const header2 = renderer.create(<Header />).toJSON();
    expect(header2).toMatchSnapshot();
});

it('renders navigation and user info when logged in', () => {
    const header = shallow(<Header isAuthenticated={true} />);
    expect(header.find(HeaderNav).length).toBe(1);
    expect(header.find(HeaderUser).length).toBe(1);
});
