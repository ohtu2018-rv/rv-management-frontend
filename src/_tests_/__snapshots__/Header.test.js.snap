// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`does not render navigation and user info when not logged in 1`] = `
<header
  className="topheader"
>
  <div
    className="header-title"
  >
    <h1>
      RV management
    </h1>
  </div>
</header>
`;

exports[`renders without crashing 1`] = `
<header
  className="topheader"
>
  <div
    className="header-title"
  >
    <h1>
      RV management
    </h1>
  </div>
</header>
`;
